network/message_scope.go

```diff
@@ -0,0 +1,206 @@
+package network
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/onflow/flow-go/crypto/hash"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/network/channels"
+	"github.com/onflow/flow-go/network/message"
+)
+
+const (
+	// eventIDPackingPrefix is used as a salt to generate payload hash for messages.
+	eventIDPackingPrefix = "libp2ppacking"
+)
+
+// ProtocolType defines the type of the protocol a message is sent over. Currently, we have two types of protocols:
+// - unicast: a message is sent to a single node through a direct connection.
+// - pubsub: a message is sent to a set of nodes through a pubsub channel.
+type ProtocolType string
+
+func (m ProtocolType) String() string {
+	return string(m)
+}
+
+const (
+	// ProtocolTypeUnicast is protocol type for unicast messages.
+	ProtocolTypeUnicast ProtocolType = "unicast"
+
+	// ProtocolTypePubSub is the protocol type for pubsub messages.
+	ProtocolTypePubSub ProtocolType = "pubsub"
+)
+
+// IncomingMessageScope captures the context around an incoming message that is received by the network layer.
+type IncomingMessageScope struct {
+	originId       flow.Identifier     // the origin node ID.
+	targetIds      flow.IdentifierList // the target node IDs (i.e., intended recipients).
+	eventId        hash.Hash           // hash of the payload and channel.
+	msg            *message.Message    // the raw message received.
+	decodedPayload interface{}         // decoded payload of the message.
+	protocol       ProtocolType        // the type of protocol used to receive the message.
+}
+
+func NewIncomingScope(originId flow.Identifier, protocol ProtocolType, msg *message.Message, decodedPayload interface{}) (*IncomingMessageScope, error) {
```

Comment:
please document the return errors

Reply:
https://github.com/onflow/flow-go/pull/3675/commits/f6d39a051221eedf3b2bfc42a90ff21d9a546369